make.js

```diff
@@ -327,32 +330,74 @@ target.test = function() {
     mkdir('-p', path.join(buildTestsPath, 'lib'));
     matchCopy(path.join('**', '@(*.ps1|*.psm1)'), path.join(__dirname, 'Tests', 'lib'), path.join(buildTestsPath, 'lib'));
 
-    // find the tests
     var suiteType = options.suite || 'L0';
-    var taskType = options.task || '*';
-    var pattern1 = buildPath + '/' + taskType + '/Tests/' + suiteType + '.js';
-    var pattern2 = buildPath + '/Common/' + taskType + '/Tests/' + suiteType + '.js';
-    var pattern3 = buildTestsPath + '/' + suiteType + '.js';
+    function runTaskTests(taskName, failWithNoTests) {
+        banner('Testing: ' + taskName);
+        // find the tests
+        var nodeVersion = options.node || getTaskNodeVersion(buildPath, taskName) + "";
+        var pattern1 = buildPath + '/' + taskName + '/Tests/' + suiteType + '.js';
+        var pattern2 = buildPath + '/Common/' + taskName + '/Tests/' + suiteType + '.js';
 
-    var testsSpec = [];
+        var testsSpec = [];
 
-    if (matchFind(pattern1, buildPath).length > 0) {
-	testsSpec.push(pattern1);
-    }
-    if (matchFind(pattern2, buildPath).length > 0) {
-	testsSpec.push(pattern2);
-    }
+        if (fs.existsSync(pattern1)) {
+            testsSpec.push(pattern1);
+        }
+        if (fs.existsSync(pattern2)) {
+            testsSpec.push(pattern2);
+        }
 
-    testsSpec = testsSpec.concat(matchFind(pattern3, buildTestsPath, { noRecurse: true }));
+        if (!testsSpec.length && !process.env.TF_BUILD) {
+            if (failWithNoTests) {
+                fail(`Unable to find tests using the following patterns: ${JSON.stringify([pattern1, pattern2])}`);
+            } else {
+                console.warn(`Unable to find tests using the following patterns: ${JSON.stringify([pattern1, pattern2])}`);
+            }
+            return;
+        }
+        // setup the version of node to run the tests
+        util.installNode(nodeVersion);
```

Comment:
I think we need to go a different route here. Right now, we iterate through all the tasks in order, for each one check the version of node needed, and then download node. This means, depending on where we are in the conversion process, we could be alternating back and forth between downloading Node 6, 10, and 14.

It would be much better IMO to first gather up the tasks we need to process for each Node version and run them clumped together by handler (e.g. all the Node 6 ones, then the Node 10 ones, then the Node 14 ones)

Reply:
We would only download once per version, so we would just be appending the PATH variable (although now I realize this will append, not change, the PATH variable). 

The problem is a bit harder when we consider that some tasks don't have a Node version (just powershell). I'll try this out and see if I can get it to work.

Reply:
Ah, missed that we [handle caching it](https://github.com/microsoft/azure-pipelines-tasks/blob/95d20693d9defd072bb5af4cce8a85b6fe7a9a11/make-util.js#L449). That's good, I was disappointed when I originally thought we didn't - I care much less about this then.

Reply:
The only real way to accomplish this is to move tasks into sub folders in the "_build" folder. Crafting a mocha command with a list of all possible tasks is too long. The method of running tests for individual tasks is okay for now.